Signing SDK Updates

derekpierre opened the pull request:
**Type of PR:**

- [x] Bugfix
- [ ] Feature
- [ ] Documentation
- [x] Other

**Required reviews:**

- [ ] 1
- [x] 2
- [ ] 3

**What this does:**

Based over #677 

- Refactor `PorterClient` to focus solely on communicating with the Porter instance i.e. sending input and returning output, without performing additional response processing. It should only handle marshalling and unmarshalling data, leaving all logic beyond that to the caller.
- Shift all post-processing logic from `PorterClient` to the `sign` module to streamline `PorterClient` responsibilities.
- Eliminate the `SigningOption` concept, as signature thresholding and hash verification are always required for `signUserOp` to produce a meaningful result.
- Fix signature aggregation by properly combining signature byte arrays instead of concatenating hex strings, which previously introduced a bug that duplicated `0x` prefixes in the result.
- Clearly differentiate failure cases when an aggregated signature cannot be produced due to either excessive errors or mismatched `messageHash` values across Porter responses. While mismatched hashes are an edge case (potentially due to malicious nodes), they are explicitly handled. In either scenario, the `sign` API will raise an error.

**Issues fixed/closed:**
Follow-up to #664

derekpierre committed:
Simplify Porter functionality to focus only on input/output to/from Porter instance without additional logic for further processing response from Porter.
The PorterClient should only focus on what Porter wants and marshalling what Porter returns for the caller to then perform their own processing.

derekpierre committed:
Move processing logic out of Porter and into sign module to simplify PorterClient code.
Remove concept of SigningOption since we'll always need to verify matching hashes, threshold of signatures to return anything meaningful from `signUserOp`.
Properly aggregate signatures returned by Porter using bytes instead of just hex strings; fixed bug where 0x was repeated within aggregated signature because hex strings were simply appended to each other.
Delineate error cases when aggregated signature cannot be produced because of too many errors or too many mismatched hashes based on responses from Porter. If an aggregated signature cannot be produced the function will throw an error.

derekpierre committed:
Improve signing integration test to ensure that aggregated signature is properly produced based on responses from Porter.

derekpierre committed:
Update porter and sign module tests to no longer use SigningOptions, and to use proper hex for signatures.
Since logic moved from Porter to the sign module, some tests were moved over to the sign test module as well.

derekpierre committed:
Improve logic for determining errors vs mismatched hashes. Mismatched hashes is a real edge case, but we handle it nonetheless.

derekpierre committed:
No need to check whether signingResults or errors are defined or not.
Add console error message for mismatched hashes.
Code cleanup.

derekpierre commented on packages/taco/src/sign.ts:
```diff
@@ -1,18 +1,54 @@
+function aggregateSignatures(signaturesByAddress: {
+  [checksumAddress: string]: TacoSignature;
+}): string {
+  // Aggregate hex signatures by concatenating them; being carefule to remove the '0x' prefix from each signature except the first one.
```
✔️

manumonti commented on packages/taco/integration-test/sign.test.ts:
```diff
@@ -9,7 +15,7 @@ import { signUserOp } from '../src/sign';
 const DUMMY_ADDRESS = '0x742D35Cc6634C0532925A3b8D33c9c0E7B66C8E8';
 const DOMAIN = 'lynx';
-const RITUAL_ID = 1;
+const COHORT_ID = 1;
```
out of the scope of this PR, but this is interesting. I have always thought that `ritual_id` is not a good name for the cohorts. I like cohort_id. Should we use this name from now on in all places?

derekpierre replied:
Interesting thought and something we can collectively discuss. Ritual ID has been ingrained into the `Coordinator` contract side of things, but doesn't mean we can't update it.
(cc @arjunhassard , @cygnusv , @vzotova ).

manumonti reviewed (approved):
LGTM! 👌

derekpierre committed:
Remove errors from SignResult since if there aren't sufficient signatures, and error is thrown.
Aggregated signatures only uses the threshold of signatures provided to it.
Add signUserOp method documentation.

derekpierre commented:
Regarding the removal of `SigningOptions`, I spoke with both @theref and @cygnusv individually. While we’ll likely end up re-adding `SigningOptions`, there’s still a lot to unpack around how the various features should be implemented, for example, supporting a non-optimistic setup that waits for a buffered set of responses beyond the first threshold, incorporating additional validations, etc. 

Given the depth of this topic and the potential logic associated with it, let's tackle it as a separate issue/PR (I'll create one soon) to encourage focused discussion and a more comprehensive implementation plan. This may also have implications for the decryption flow, which currently assumes an entirely optimistic response model.
